src/bin/pg_autoctl/monitor.c

```diff
@@ -104,14 +248,77 @@ monitor_get_other_node(Monitor *monitor, char *myHost, int myPort, NodeAddress *
 
 	if (!parseContext.parsedOK)
 	{
-		log_error("Failed to get the secondary from the monitor while running "
+		log_error("Failed to get the other nodes from the monitor while running "
 				  "\"%s\" with host %s and port %d because it returned an "
 				  "unexpected result. See previous line for details.",
 				  sql, myHost, myPort);
 		return false;
 	}
 
-	log_info("Other node in the HA group is %s:%d", node->host, node->port);
+	return true;
+}
+
+
+/*
+ * monitor_get_other_node gets the hostname and port of the other node
+ * in the group.
+ */
+bool
+monitor_get_other_nodes_as_json(Monitor *monitor,
+								char *myHost, int myPort,
+								NodeState currentState,
+								char *json, int size)
+{
+	PGSQL *pgsql = &monitor->pgsql;
+	SingleValueResultContext context = { { 0 }, PGSQL_RESULT_STRING, false };
+
+	const char *sql =
+		currentState == ANY_STATE
+		?
+		"SELECT jsonb_pretty(jsonb_agg(row_to_json(nodes)))"
+		" FROM pgautofailover.get_other_nodes($1, $2) as nodes"
+		:
+		"SELECT jsonb_pretty(jsonb_agg(row_to_json(nodes)))"
+		" FROM pgautofailover.get_other_nodes($1, $2, "
+		"$3::pgautofailover.replication_state) as nodes";
+
+	int paramCount = 2;
+	Oid paramTypes[3] = { TEXTOID, INT4OID, TEXTOID };
+	const char *paramValues[3] = { 0 };
+	IntString myPortString = intToString(myPort);
+
+	paramValues[0] = myHost;
+	paramValues[1] = myPortString.strValue;
+
+	if (currentState != ANY_STATE)
+	{
+		++paramCount;
+		paramValues[2] = NodeStateToString(currentState);
+	}
+
+	if (!pgsql_execute_with_params(pgsql, sql,
+								   paramCount, paramTypes, paramValues,
+								   &context, &parseSingleValueResult))
+	{
+		log_error("Failed to get the other nodes from the monitor while running "
+				  "\"%s\" with host %s and port %d", sql, myHost, myPort);
+		return false;
+	}
+
+	/* disconnect from PostgreSQL now */
+	pgsql_finish(&monitor->pgsql);
+
+	if (!context.parsedOk)
+	{
+		log_error("Failed to get the other nodes from the monitor while running "
+				  "\"%s\" with host %s and port %d because it returned an "
+				  "unexpected result. See previous line for details.",
+				  sql, myHost, myPort);
+		return false;
+	}
+
+	strlcpy(json, context.strVal, size);
```

Comment:
context.strVal is allocated inside  `parseSingleValueResult` via `strdup` but never deallocated

Reply:
I added a call to `free` the memory there, and I noticed other places that share the same problem. Another PR to fix those places would be good.